onnxruntime/core/providers/coreml/builders/impl/argmax_op_builder.cc

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/shared/utils/utils.h"
+#include "core/providers/coreml/builders/model_builder.h"
+#include "core/providers/coreml/builders/op_builder_factory.h"
+
+#include "base_op_builder.h"
+
+namespace onnxruntime {
+namespace coreml {
+
+class ArgMaxOpBuilder : public BaseOpBuilder {
+  // Add operator related
+ private:
+  Status AddToModelBuilderImpl(ModelBuilder& model_builder, const Node& node,
+                               const logging::Logger& logger) const override ORT_MUST_USE_RESULT;
+
+  // Operator support related
+ private:
+  bool IsOpSupportedImpl(const Node& node, OpBuilderInputParams& input_params,
+                         const logging::Logger& logger) const override;
+};
+
+// Add operator related
+
+Status ArgMaxOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder,
+                                              const Node& node,
+                                              const logging::Logger& /* logger */) const {
+  std::unique_ptr<COREML_SPEC::NeuralNetworkLayer> layer = CreateNNLayer(node);
+  const auto& graph_viewer = model_builder.GetGraphViewer();
+
+  NodeAttrHelper helper(node);
+  const auto axis = helper.Get("axis", 0);
+  const auto keep_dims = helper.Get("keep_dims", 1);
+  bool removedim = false;
+  if (keep_dims != 1) {
+    removedim = true;
+  }
+
+  auto* coreml_argmax = layer->mutable_argmax();
+  coreml_argmax->set_axis(axis);
+  coreml_argmax->set_removedim(removedim);
+
+  // Get ArgMax's next node(Cast)'s outputdefs
+  auto it = node.OutputEdgesBegin();
+  const auto* succ_node(graph_viewer.GetNode(it->GetNode().Index()));
+
+  // Skip the cast's input/argmax's output
+  *layer->mutable_input()->Add() = node.InputDefs()[0]->Name();
+  *layer->mutable_output()->Add() = succ_node->OutputDefs()[0]->Name();
+
+  model_builder.AddLayer(std::move(layer));
+  return Status::OK();
+}
+
+// Operator support related
+
+bool ArgMaxOpBuilder::IsOpSupportedImpl(const Node& node, OpBuilderInputParams& input_params,
+                                        const logging::Logger& logger) const {
+  // Check if Argmax's output is the graph output
+  const auto& graph_output_list = input_params.graph_viewer.GetOutputs();
+  std::unordered_set<const NodeArg*> graph_outputs(graph_output_list.cbegin(), graph_output_list.cend());
+  const auto& output_defs = node.OutputDefs();
+  for (const auto* output_def : output_defs) {
+    if (graph_outputs.count(output_def) != 0) {
+      LOGS(logger, VERBOSE) << "Case: ArgMax's output is the graph output: Not supported.";
+      return false;
```

Comment:
nit: error message is for user and they don't care about the 'case: ' part. I'd suggest something simple like 'ArgMax not supported when it produces a graph output'.

What's the reason it can't use CoreML if it's producing a graph output though? How is that different to producing a value used by another node?

Reply:
I think the reason is not producing a graph output is so different from producing a node output. Right now coreml ep handles only float/int32 graph input/output type, we certainly can add more type in the future. 
But for this specific argmax op, if it's the last node in the graph and its output is the graph output, meaning that it's an int64 type which is just not supported in coreml model.  In other words, for us the output should be int64, but coreml only supports int32 output. That's also the reason why for cast node if it's the graph output, we need to make sure cast_to type is int32 too, ( to make sure it's a supported output type in coreml.)

Reply:
The graph output type error comes here: [https://github.com/microsoft/onnxruntime/pull/7924/files#diff-bcd08e99c1ac1f35f26c01a98a87e4795aaf16c6f0004a829260a7f1c494e468R160](url)

Reply:
And the graph output type error is thrown here: (If without the exit for the case argmax is the graph output)[https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/core/providers/coreml/builders/model_builder.cc#L156-L158](url)